### PR TITLE
Add support for U32_TYPE and NUMBER_LITERAL

### PR DESCRIPTION
**Pull request by kpadmasola**

### CODE REVIEW

File: parser.cpp [hunk 1]
```diff
@@ -193,24 +197,18 @@ class Parser {
                 if (*(str + 1) == '3' && *(str + 2) == '2') {
                     advance_pos(data, 3);
 
-                    if (parse_rest_of_identifier(data) == token::IDENTIFIER) {
-                        return token::IDENTIFIER;
-                    }
-
-                    // if next char is not part of an identifier, we found a u32 type
-                    return token::U32_TYPE;
+                    if (!(isalnum(**data) || **data == '_'))
```
Comment (luc-tielen) [inline parser.cpp]: Make a helper function for this: `is_identifier_char`.

File: parser.cpp [hunk 2]
```diff
@@ -193,24 +197,18 @@ class Parser {
...
                 if (memcmp (str + 1, "tring", 5) == 0) {
                     advance_pos(data, 6);
 
-                    if (parse_rest_of_identifier(data) == token::IDENTIFIER) {
-                        return token::IDENTIFIER;
-                    }
-
-                    // if next char is not part of an identifier, we found a u32 type
-                    return token::STRING_TYPE;
+                    if (!(isalnum(**data) || **data == '_'))
```
Comment (luc-tielen) [inline parser.cpp]: use the helper function here too

File: parser.cpp (hunk 1 above)
Comment (luc-tielen) [reply parser.cpp]: Also: Was there something that didn't parse if we didn't switch the condition around?

File: parser.cpp [hunk 3]
```diff
@@ -238,6 +236,26 @@ class Parser {
...
     }
 
+    Token parse_number_literal(const char** data) {
+        auto str = *data;
+        while (is_not_at_eof()) {
+            auto c = *str;
+            if (isdigit(c)) {
+                str++;
+                pos++;
+            } else if (c == '(' || c == ',' || c == ')' || c == '.') {
```
Comment (luc-tielen) [inline parser.cpp]: I think a check for whitespace is missing here.

Comment (luc-tielen) [reply parser.cpp]: Also does this set the next position right before any of these characters? (For correct positions of later tokens?)

**Review by luc-tielen — CHANGES_REQUESTED** [review]

File: parser.cpp (hunk 1 above)
Comment (kpadmasola) [reply parser.cpp]: yes, U32 was getting parsed as IDENTIFIER earlier

Comment (luc-tielen) [reply parser.cpp]: Oops. Thanks for fixing this!

File: parser.cpp (hunk 3 above)
Comment (kpadmasola) [reply parser.cpp]: Yes, when it comes out of the function, `pos` will be pointing to one of these characters. Or to put it another way, `pos` in only incremented for a digit.

**Review by luc-tielen — APPROVED** [review]

**Pull request merged**